src/core/server/http/router/validator/validator.ts

```diff
@@ -0,0 +1,244 @@
+/*
+ * Licensed to Elasticsearch B.V. under one or more contributor
+ * license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright
+ * ownership. Elasticsearch B.V. licenses this file to you under
+ * the Apache License, Version 2.0 (the "License"); you may
+ * not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+import { ValidationError, Type, schema, ObjectType } from '@kbn/config-schema';
+import { Stream } from 'stream';
+import { RouteValidationError } from './validator_error';
+
+/**
+ * Allowed returned format of the custom validate function
+ * @public
+ */
+export type RouteValidateFunctionReturn<T> =
+  | {
+      value: T;
+      error?: never;
+    }
+  | {
+      value?: never;
+      error: RouteValidationError;
+    };
```

Comment:
I think the intellisense would be improved if this was inlined in the `RouteValidateFunction` type. It's only referenced in one other place which can be easily replaced with a `ReturnType` generic.